File: empyrean-py/python/empyrean/observers/observers.py
"""Observer state type."""

from collections.abc import Sequence

import numpy as np
import quivr as qv

from empyrean._convert import frame_to_int, int_to_frame, naif_to_origin, origin_to_naif
from empyrean.coordinates.coordinates import CartesianCoordinates
from empyrean.coordinates.enums import Frame, Origin
from empyrean.coordinates.epoch import Epochs


class Observers(qv.Table):
    """Precomputed observer states as Cartesian coordinates.

    Each row is one MPC observatory at one epoch. The kinematic state
    lives inside a nested :class:`CartesianCoordinates` table — same
    schema as orbit positions — so frame and origin are explicit and
    consistent with the rest of the API. By default that's
    :attr:`Frame.ICRF` and :attr:`Origin.SSB` — the **construction
    basis**, and the one every downstream consumer (ephemeris
    generation, orbit determination) requires. Pass ``frame`` /
    ``origin`` to :meth:`from_code` / :meth:`from_codes` to get the
    states in another basis instead, for geometry work that wants one
    (e.g. heliocentric ecliptic site positions).

    Construct via :meth:`from_code` (single observatory, ``N`` epochs)
    or :meth:`from_codes` (cartesian product of ``N`` observatories ×
    ``M`` epochs). The top-level shortcut
    :func:`empyrean.get_observer_states` is the same as
    :meth:`from_codes`.
    """

    obs_code = qv.LargeStringColumn()
    coordinates = CartesianCoordinates.as_column()
    observing_night = qv.Int32Column(nullable=True)  # YYYYMMDD or null

    @classmethod
    def from_code(
        cls,
        obs_code: str,
        epochs: Epochs | np.ndarray | Sequence[float],
        frame: Frame | str | int = Frame.ICRF,
        origin: Origin | str | int = Origin.SSB,
    ) -> "Observers":
        """Observer states for a **single** observatory at ``N`` epochs.

        Returns an ``N``-row table, one row per epoch.

        Parameters
        ----------
        obs_code : str
            MPC observatory code (e.g. ``"W84"``, ``"F51"``, ``"274"``).
        epochs : Epochs | array-like
            ``N`` observation epochs. An :class:`Epochs` table is
            converted to TDB internally; an array is treated as MJD TDB.
        frame : Frame | str | int
            Reference frame for the returned states. Default
            :attr:`Frame.ICRF`. See :meth:`from_codes`.
        origin : Origin | str | int
            Body the states are relative to. Default
            :attr:`Origin.SSB`. See :meth:`from_codes`.

        Returns
        -------
        Observers
            Length-``N`` table — row ``i`` is the observer at ``epochs[i]``.

        Examples
        --------
        >>> times = Epochs.from_mjd([60500.0, 60501.0])
        >>> obs = Observers.from_code("W84", times)
        >>> len(obs)
        2
        """
        return cls.from_codes([obs_code], epochs, frame=frame, origin=origin)

    @classmethod
    def from_codes(
        cls,
        obs_codes: Sequence[str],
        epochs: Epochs | np.ndarray | Sequence[float],
        frame: Frame | str | int = Frame.ICRF,
        origin: Origin | str | int = Origin.SSB,
    ) -> "Observers":
        """Observer states for the **cartesian product** of ``N``
        observatory codes × ``M`` epochs.

        Returns an ``N * M``-row table in **code-major** order: all
        ``M`` epochs for ``obs_codes[0]``, then all ``M`` epochs for
        ``obs_codes[1]``, and so on. Indexing into the result is
        therefore ``i * M + j`` for the (code ``i``, epoch ``j``) row.

        Parameters
        ----------
        obs_codes : list[str]
            ``N`` MPC observatory codes.
        epochs : Epochs | array-like
            ``M`` observation epochs. An :class:`Epochs` table is
            converted to TDB internally; an array is treated as MJD TDB.
        frame : Frame | str | int
            Reference frame for the returned states. Default
            :attr:`Frame.ICRF`.
        origin : Origin | str | int
            Body the states are relative to. Default :attr:`Origin.SSB`.

        Notes
        -----
        ``(Frame.ICRF, Origin.SSB)`` is the **construction basis** — the
        one ephemeris generation and orbit determination require, and the
        one to keep when the observers are headed there. Requesting it
        takes no transform at all: the states come back exactly as
        constructed, bit for bit. Any other basis rotates and/or
        translates them, which is what a consumer plotting observer
        geometry wants (e.g. heliocentric ecliptic site positions via
        ``frame=Frame.ECLIPTICJ2000, origin=Origin.SUN``).

        The ``frame`` / ``origin`` stamped on the returned table are read
        off the states the engine returned, not echoed from the request,
        so a table always reports the basis that actually produced it.

        Returns
        -------
        Observers
            ``N * M``-row table, code-major.

        Examples
        --------
        >>> times = Epochs.from_mjd([60000.0, 60001.0])
        >>> obs = Observers.from_codes(["W84", "F51", "274"], times)
        >>> len(obs)
        6
        >>> obs.obs_code.to_pylist()
        ['W84', 'W84', 'F51', 'F51', '274', '274']
        """
        from empyrean._empyrean_rs import _get_observers

        if isinstance(epochs, Epochs):
            tdb = epochs.to_tdb()
            epochs_mjd = np.asarray(tdb.mjd.to_numpy(zero_copy_only=False), dtype=np.float64)
        else:
            epochs_mjd = np.asarray(epochs, dtype=np.float64)

        result = _get_observers(
            list(obs_codes), epochs_mjd, frame_to_int(frame), origin_to_naif(origin)
        )

        nights = result["observing_night"]
        night_list = [int(n) if n >= 0 else None for n in nights]

        # Basis as reported by the returned states, never echoed from the
        # request. `frame` is a table-level attribute in quivr, so a
        # mixed-frame return would be unrepresentable — surface it loudly
        # instead of silently keeping the first row's value.
        frames = np.asarray(result["frame"])
        origins = np.asarray(result["origin"])
        if len(frames) == 0:
            out_frame = int_to_frame(frame_to_int(frame))
        elif np.all(frames == frames[0]):
            out_frame = int_to_frame(int(frames[0]))
        else:
            raise ValueError(
                "observer states came back in more than one frame "
                f"({sorted({int(f) for f in frames})}); Observers carries a single "
                "table-level frame and cannot represent a mixed-frame table"
            )

        coordinates = CartesianCoordinates.from_kwargs(
            epoch=np.asarray(result["epoch"]),
            x=np.asarray(result["x"]),
            y=np.asarray(result["y"]),
            z=np.asarray(result["z"]),
            vx=np.asarray(result["vx"]),
            vy=np.asarray(result["vy"]),
            vz=np.asarray(result["vz"]),
            frame=out_frame.value,
            origin=[naif_to_origin(int(o)) for o in origins],
        )

        return cls.from_kwargs(
            obs_code=result["obs_code"],
            coordinates=coordinates,
            observing_night=night_list,
        )

File: empyrean/src/observers.rs
//! MPC observatory code → observer state queries.

use crate::context::Context;
use crate::coordinate::{Frame, Origin, int_to_frame};
use crate::error::{Error, Result};
use std::ffi::CString;

/// Observer state at one epoch.
#[derive(Debug, Clone, PartialEq)]
pub struct Observer {
    /// MPC observatory code (3 characters).
    pub obs_code: String,
    /// Epoch.
    pub epoch: crate::Epoch,
    /// Position in the [`frame`](Self::frame) / [`origin`](Self::origin)
    /// basis (AU).
    pub position: [f64; 3],
    /// Velocity in the [`frame`](Self::frame) / [`origin`](Self::origin)
    /// basis (AU/day).
    pub velocity: [f64; 3],
    /// Observing night as YYYYMMDD integer, or -1 if unavailable
    /// (space-based observers, or a site without a longitude).
    ///
    /// The night is the local calendar date on which the observing night
    /// began: the UTC epoch is shifted by the site's east longitude to
    /// local mean solar time, and epochs before local noon stamp the
    /// previous date. MPC east-of-Greenwich longitudes in \\( [0, 360) \\)
    /// are wrapped to signed \\( [-180, 180] \\) before the fold, so a
    /// site west of Greenwich (e.g. \\( 289^\circ \\)E in Chile) shifts by
    /// \\( -71^\circ \\), not \\( +19 \\) hours.
    pub observing_night: i32,
    /// Reference frame [`position`](Self::position) and
    /// [`velocity`](Self::velocity) are expressed in.
    ///
    /// Read off the returned state rather than echoed from the request,
    /// so a row always reports the basis that actually produced it.
    pub frame: Frame,
    /// Body the state is relative to.
    pub origin: Origin,
}

impl Observer {
    pub(crate) fn from_ffi(o: &empyrean_sys::EmpyreanObserver) -> Result<Self> {
        Ok(Self {
            obs_code: obs_code_from_bytes(&o.obs_code),
            epoch: crate::Epoch::from_mjd_tdb(o.epoch_mjd_tdb),
            position: [o.x, o.y, o.z],
            velocity: [o.vx, o.vy, o.vz],
            observing_night: o.observing_night,
            frame: int_to_frame(o.frame)?,
            origin: Origin::from_naif_id(o.origin).ok_or_else(|| {
                Error::invalid_input(format!(
                    "libempyrean returned an unknown origin NAIF id: {}",
                    o.origin
                ))
            })?,
        })
    }
}

pub(crate) fn obs_code_from_bytes(bytes: &[u8; 4]) -> String {
    let end = bytes.iter().position(|&b| b == 0).unwrap_or(bytes.len());
    String::from_utf8_lossy(&bytes[..end]).into_owned()
}

impl Context {
    /// Compute observer states (cross product: codes × epochs) in a
    /// caller-chosen `(frame, origin)` basis.
    ///
    /// Returns `obs_codes.len() * epochs.len()` observer entries in
    /// code-major order: all epochs for `obs_codes[0]`, then all epochs
    /// for `obs_codes[1]`, so `result[i * epochs.len() + j]` is
    /// `(obs_codes[i], epochs[j])`.
    ///
    /// # Choosing a basis
    ///
    /// [`Frame::ICRF`] with [`Origin::SSB`] is the **construction
    /// basis** — the one every consumer of an observer state requires,
    /// and the one to pass when the observers are headed for ephemeris
    /// generation or orbit determination. Requesting it takes no
    /// transform at all: the observers come back exactly as constructed,
    /// bit for bit. Any other basis rotates and/or translates them, which
    /// is what a consumer plotting observer geometry wants (e.g.
    /// heliocentric ecliptic site positions via
    /// `(Frame::EclipticJ2000, Origin::SUN)`).
    ///
    /// # Errors
    ///
    /// Every rejection aborts the whole call — there is no partial output
    /// and no silently untransformed entry. The failures keep the
    /// engine's two axes rather than being flattened onto one:
    /// [`Error::code`] is `-1` when the remedy is different arguments (an
    /// unknown observatory code, a frame an observer cannot be rotated
    /// into, an origin that is an MPC site rather than an SPK body) and
    /// `-2` when the remedy is to load or fetch data (an unfetched
    /// spacecraft kernel, an SPK coverage gap, a BPC window that ends
    /// before the requested epoch).
    pub fn get_observers(
        &self,
        obs_codes: &[&str],
        epochs: &[crate::Epoch],
        frame: Frame,
        origin: Origin,
    ) -> Result<Vec<Observer>> {
        let cstrings: Vec<CString> = obs_codes
            .iter()
            .map(|&c| {
                CString::new(c).map_err(|_| Error::invalid_input("observatory code has NUL byte"))
            })
            .collect::<Result<Vec<_>>>()?;
        let ptrs: Vec<*const std::ffi::c_char> = cstrings.iter().map(|s| s.as_ptr()).collect();
        let epochs_mjd_tdb: Vec<f64> = epochs
            .iter()
            .map(|e| e.mjd_tdb())
            .collect::<Result<Vec<_>>>()?;

        let mut result = empyrean_sys::EmpyreanObserverResult {
            observers: std::ptr::null_mut(),
            num_observers: 0,
        };
        let code = unsafe {
            empyrean_sys::empyrean_get_observers(
                self.as_raw(),
                ptrs.as_ptr(),
                ptrs.len(),
                epochs_mjd_tdb.as_ptr(),
                epochs_mjd_tdb.len(),
                frame as i32,
                origin.naif_id(),
                &mut result,
            )
        };
        if code != 0 {
            return Err(Error::capture(code));
        }
        let observers: Result<Vec<Observer>> = unsafe {
            std::slice::from_raw_parts(result.observers, result.num_observers)
                .iter()
                .map(Observer::from_ffi)
                .collect()
        };
        // Free before propagating: the native array is owned regardless
        // of whether marshaling every row succeeded.
        unsafe { empyrean_sys::empyrean_observer_result_free(&mut result) };
        observers
    }
}
